### PR TITLE
New module ec2_asg_facts

### PR DESCRIPTION
**Pull request by manicminer**
##### Issue Type:
- New Module Pull Request
##### Plugin Name:

ec2_asg_facts
##### Ansible Version:

```
ansible 2.1.0
  config file = /path/to/my/ansible.cfg
  configured module search path = Default w/o overrides
```
##### Summary:

Searches AWS auto scaling groups in a region and returns fact information. Can return details of all auto scaling groups, a subset, or a single one, identified by a regular expression match against the name, or a list of matching tags.

This PR replaces https://github.com/ansible/ansible-modules-extras/pull/1351 which in turn replaces https://github.com/ansible/ansible-modules-extras/pull/1659.

Authors: @linuxdynasty and myself
##### Example output:

```
TASK [Retrieve ASG details] ****************************************************
changed: [localhost -> localhost] => {"changed": true, "rc": 0, "results": [{"auto_scaling_group_arn": "arn:aws:autoscaling:us-west-2:123456789012:autoScalingGroup:6129a1fd-223a-837a-8fbe-04d1227def03:autoScalingGroupName/my-group-20160221202916", "auto_scaling_group_name": "my-group-20160221202916", "availability_zones": ["us-west-2c", "us-west-2b", "us-west-2a"], "created_time": "2016-02-21T20:29:19.021000+00:00", "default_cooldown": 300, "desired_capacity": 2, "enabled_metrics": [], "health_check_grace_period": 300, "health_check_type": "EC2", "instances": [{"availability_zone": "us-west-2c", "health_status": "Healthy", "instance_id": "i-4a74c2ae", "launch_configuration_name": "my-group-lc-20160221203934-master-317daac", "lifecycle_state": "InService", "protected_from_scale_in": false}, {"availability_zone": "us-west-2a", "health_status": "Healthy", "instance_id": "i-3613bcdc", "launch_configuration_name": "my-group-lc-20160221203934-master-317daac", "lifecycle_state": "InService", "protected_from_scale_in": false}], "launch_configuration_name": "my-group-lc-20160221203934-master-317daac", "load_balancer_names": ["my-elb-1", "my-elb-2"], "max_size": 4, "min_size": 2, "new_instances_protected_from_scale_in": false, "placement_group_status": null, "suspended_processes": [], "tags": [{"key": "Name", "propagate_at_launch": true, "resource_id": "my-group-20160221202916", "resource_type": "auto-scaling-group", "value": "my-instance-20160221202916"}, {"key": "environment", "propagate_at_launch": true, "resource_id": "my-group-20160221202916", "resource_type": "auto-scaling-group", "value": "staging"}, {"key": "project", "propagate_at_launch": true, "resource_id": "my-group-20160221202916", "resource_type": "auto-scaling-group", "value": "my-project"}, {"key": "role", "propagate_at_launch": true, "resource_id": "my-group-20160221202916", "resource_type": "auto-scaling-group", "value": "web"}], "termination_policies": ["Default"], "vpc_zone_identifier": ["subnet-359c64e3", "subnet-a18f77ed", "subnet-0742e84a"]}]}
```

### CODE REVIEW

**Comment by sivel** [comment]
```
============================================================================
./cloud/amazon/ec2_asg_facts.py
============================================================================
ERROR: indentation contains tabs. line 314 column 0
```

**Comment by manicminer** [comment]
Oops, fixed and pushed.

**Comment by linuxdynasty** [comment]
@tombamford did you close out the last PR for this one instead?

**Comment by manicminer** [comment]
@linuxdynasty I wasn't able to close the other PR, probably due to permissions as I hadn't opened it. I needed to rebase and push a new branch so figured it would probably be easier to open a new PR.

**Comment by manicminer** [comment]
Pinging @gregdek, @wimnat for review as you have seen the previous two requests for inclusion. Thanks!

**Comment by robynbergeron** [comment]
Thanks @tombamford for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

**Comment by manicminer** [comment]
@gregdek @wimnat Would you perhaps be able to review this? Asking as you already looked at the previous PRs for this change (https://github.com/ansible/ansible-modules-extras/pull/1659 and https://github.com/ansible/ansible-modules-extras/pull/1351). Thanks!

File: cloud/amazon/ec2_asg_facts.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+        if name:
+            name_prog = re.compile(r'^' + name)
+            matched_name = name_prog.search(asg['AutoScalingGroupName'])
+        else: matched_name = True
+
+        if tags: matched_tags = match_asg_tags(tags, asg)
+        else: matched_tags = True
+
+        if matched_name and matched_tags:
+            asg_details = dict(created_time=asg['CreatedTime'].isoformat())
```
Comment (wimnat) [inline cloud/amazon/ec2_asg_facts.py:310]: Assuming speedy merge of https://github.com/ansible/ansible/pull/15200, you can replace everything in this if block with:

```
asg['CreatedTime'] = asg['CreatedTime'].isoformat()
asg_details = camel_dict_to_snake_dict(asg)
```

File: cloud/amazon/ec2_asg_facts.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+            }
+        ]
+    """
+    asgs = conn.describe_auto_scaling_groups()['AutoScalingGroups']
+    matched_asgs = []
+    for asg in asgs:
+        if name:
+            name_prog = re.compile(r'^' + name)
+            matched_name = name_prog.search(asg['AutoScalingGroupName'])
+        else: matched_name = True
```
Comment (wimnat) [inline cloud/amazon/ec2_asg_facts.py:304]: pep8 - mutliple statements on one line

File: cloud/amazon/ec2_asg_facts.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+
+        if tags: matched_tags = match_asg_tags(tags, asg)
```
Comment (wimnat) [inline cloud/amazon/ec2_asg_facts.py:306]: pep8

File: cloud/amazon/ec2_asg_facts.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+        else: matched_tags = True
```
Comment (wimnat) [inline cloud/amazon/ec2_asg_facts.py:307]: pep8

File: cloud/amazon/ec2_asg_facts.py [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+    asg_tags = module.params.get('tags')
+
+    try:
+        region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+        autoscaling = boto3_conn(module, conn_type='client', resource='autoscaling', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+    except botocore.exceptions.ClientError, e:
+        module.fail_json(msg="Boto3 Client Error - " + str(e.msg))
+
+    results = find_asgs(autoscaling, name=asg_name, tags=asg_tags)
+    module.exit_json(results=results)
```
Comment (wimnat) [inline cloud/amazon/ec2_asg_facts.py:362]: Up for discussion about this one.  Do we use "results=" or do we use something specific to the module, e.g. "asgs="?

I've seen both across multiple modules.

**Comment by gregdek** [comment]
Thanks @tombamford for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@manicminer A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@manicminer Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@manicminer A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Closed due to contributor inactivity.

**Comment by apolloclark** [comment]
Any updates on this PR @manicminer and @gregdek  ? I'd be willing to take it up and make any necessary changes.

**Comment by gregdek** [comment]
@apolloclark feel free to create your own branch with a new PR forked from the original repo.

**Comment by apolloclark** [comment]
@gregdek Sounds good. Anything in particular this PR needs outside of PEP8 compatibility?

**Comment by gregdek** [comment]
Nothing that I can see. Be sure to link the original PR when you open yours.

**Comment by apolloclark** [comment]
Will do. I'll post something this weekend. Thanks!

**Comment by manicminer** [comment]
@apolloclark Please feel free to rebase/rework and submit this if you wish.

**Comment by apolloclark** [comment]
@manicminer awesome, thanks! It's really well written, I won't be changing much. I used this code as the basis for an "ec2_lc_facts" module too.

**Comment by apolloclark** [comment]
@manicminer and @gregdek I've created a new pull request here:
https://github.com/ansible/ansible-modules-extras/pull/2565

**Comment by gregdek** [comment]
Thanks @apolloclark (and of course to @manicminer for the excellent headstart).

**Comment by wimnat** [comment]
@apolloclark i noticed your PR was actually for AMI find rather than ASG facts and it's been closed.  Maybe this was by mistake?  Anyway, I quickly added a new PR for asg_facts if you want to review

**Comment by apolloclark** [comment]
@wimnat no worries, @manicminer wrote code for multiple Amazon AWS modules for searching. I'll update the code for the ASG one today.

**Pull request closed without merging**